#165421197 File Downloads

3N61N33R opened the pull request:
#### What does this PR do?
- enables file downloads from one endpoint

#### Description of Task to be completed?
Given an admin with access
When I request to get any file for downloads 
Then I should use the files hosted on the download endpoint

#### What are the relevant pivotal tracker stories?
[#165421197](https://www.pivotaltracker.com/story/show/165421197)

Commit:
feat(downloads): downloads endpoint

- enable file downloads from one endpoint

Commit:
feat(downloads): downloads endpoint

- enable file downloads from one endpoint

egichuri commented on api/urls.py:
```diff
@@ -126,6 +127,7 @@ def __init__(self, trailing_slash="/?"):
         name="sample-import-file",
     ),
     path("filter-values/", AvailableFilterValues.as_view(), name="available-filters"),
+    path("downloads/?P<filewanted>", Downloads.as_view(), name="Downloads"),
```
we should have the query param checked inside the view. this will give us flexibility with the endpoint. see [this](https://www.django-rest-framework.org/api-guide/requests/)

egichuri commented on api/views/assets.py:
```diff
@@ -597,3 +597,97 @@ def perform_update(self, serializer):
+                status=200,
+            )
+
+    def create_sheet(self, assets_list, filename=None):
```
the exports logic should not change. let's remove this

Commit:
feat(downloads): asset file downloads

- combines 3 endpoints into one
- adds a file folder for testing functionality

Commit:
feat(downloads): asset file downloads

- combines 3 endpoints into one
- adds a file folder for testing functionality

egichuri commented on api/views/assets.py:
```diff
@@ -433,6 +433,32 @@ def post(self, request):
+            filename = "{}_exported_assets.xlsx".format(email.split("@")[0])
+
+        # send file
+        file_path = os.path.join(settings.BASE_DIR, "files/{}".format(filename))
```
we can pass the `filename` here without the checks from [L444](https://github.com/AndelaOSP/art-backend/pull/298/files#diff-3eb66745053f6b976b9e9ca1186bfd4cR444) to [L451](https://github.com/AndelaOSP/art-backend/pull/298/files#diff-3eb66745053f6b976b9e9ca1186bfd4cR451), then use try/except/else to raise an appropriate error when the file doesn't exist

egichuri commented on api/views/assets.py:
```diff
@@ -597,3 +623,97 @@ def perform_update(self, serializer):
             )
+
+
+class Downloads(APIView):
```
let's get rid of this class

egichuri reviewed (changes requested):
good progress on this.. please exclude the generated files in your commit.  
 - files/grace.zawadi.csv
 - files/grace.zawadi_exported_assets.xlsx
 - files/admin.csv

Commit:
feat: file downloads endpoint

- deletes unused view
- deletes autogenerated files
- refactors file downloads view to use try/except/else

Commit:
feat: file downloads endpoint

- deletes unused view
- deletes autogenerated files
- refactors file downloads view to use try/except/else

egichuri commented:
this should work.. let's redirect existing usage to this endpoint and remove the old ones

Commit:
- fixes failing tests
- adjusts tests to test altered implementation

Commit:
- fixes failing tests
- adjusts tests to test altered implementation

Commit:
- test

Commit:
- test

kenneth051 committed:
user emails should be andela emails (#302)

[Delivers #164753083]

kmazi commented:
@egichuri is this good to go?

egichuri commented:
@graycadeau do we have a way to redirect existing usage?

3N61N33R commented:
> @graycadeau do we have a way to redirect existing usage?

How do you mean? @egichuri

kmazi commented:
> > @graycadeau do we have a way to redirect existing usage?
> 
> How do you mean? @egichuri

It could mean that all existing means of downloading files should be deprecated in favor of this one.

egichuri commented:
@graycadeau @kmazi - the changes are ok.. meant to ask if the existing endpoints still work. if not we can work with the FE to ensure they update the endpoints before this is merged.

RonKbS committed:
#165377092 Allow posting and getting of asset invoice receipts (#304)

* feat(asset): allow posting and getting of asset invoice receipts

- allow only a super admin to post an asset's invoice_receipt
- allow only an admin to view an asset's invoice_receipt

[Finishes #165377092]

* - make restrict_to_only_super_users an internal function
- add assertions for file_upload

[Finishes #165377092]

*  - fix black-formatting error in test_manage_asset_api file

* - add settings to allow viewing of uploaded files
- add test that shows a super_admin can patch an asset's invoice receipt

* - remove test file uploaded during patching by super-admin

kmazi commented:
> @graycadeau @kmazi - the changes are ok.. meant to ask if the existing endpoints still work. if not we can work with the FE to ensure they update the endpoints before this is merged.

@egichuri could you elaborate on what existing endpoints you're referring to. @graycadeau I'm sure you know what endpoints he's talking about.

3N61N33R committed:
#166114088 Assigner history (#303)

* feat(assigner history): track asset assigner

- adds assigner field to allocation history model

[Finishes #166114088]

* - adjust tests

* - test assigner in asset view allocation history

* - trim trailing whitespace

* #165377092 Allow posting and getting of asset invoice receipts (#304)

* feat(asset): allow posting and getting of asset invoice receipts

- allow only a super admin to post an asset's invoice_receipt
- allow only an admin to view an asset's invoice_receipt

[Finishes #165377092]

* - make restrict_to_only_super_users an internal function
- add assertions for file_upload

[Finishes #165377092]

*  - fix black-formatting error in test_manage_asset_api file

* - add settings to allow viewing of uploaded files
- add test that shows a super_admin can patch an asset's invoice receipt

* - remove test file uploaded during patching by super-admin

* feat(assigner history): track asset assigner

- adds assigner field to allocation history model

[Finishes #166114088]

* - adjust tests

* - test assigner in asset view allocation history

* - trim trailing whitespace

* fix migration conflict
i

3N61N33R commented:
> @graycadeau @kmazi - the changes are ok.. meant to ask if the existing endpoints still work. if not we can work with the FE to ensure they update the endpoints before this is merged.

@egichuri the 3 endpoints that were merged into one were deleted
`    path("asset-details/", GetPrintAssetsFile.as_view(), name="asset-details"),	`
`    path("skipped/", SkippedAssets.as_view(), name="skipped"),	`
`path( "files/sample_import_file/", SampleImportFile.as_view(), name="sample-import-file", ),`

kenneth051 committed:
Filter asset logs categorizations (#305)

* Filter asset logs categorizations

* #166114088 Assigner history (#303)

* feat(assigner history): track asset assigner

- adds assigner field to allocation history model

[Finishes #166114088]

* - adjust tests

* - test assigner in asset view allocation history

* - trim trailing whitespace

* #165377092 Allow posting and getting of asset invoice receipts (#304)

* feat(asset): allow posting and getting of asset invoice receipts

- allow only a super admin to post an asset's invoice_receipt
- allow only an admin to view an asset's invoice_receipt

[Finishes #165377092]

* - make restrict_to_only_super_users an internal function
- add assertions for file_upload

[Finishes #165377092]

*  - fix black-formatting error in test_manage_asset_api file

* - add settings to allow viewing of uploaded files
- add test that shows a super_admin can patch an asset's invoice receipt

* - remove test file uploaded during patching by super-admin

* feat(assigner history): track asset assigner

- adds assigner field to allocation history model

[Finishes #166114088]

* - adjust tests

* - test assigner in asset view allocation history

* - trim trailing whitespace

* fix migration conflict
i

* Filter asset logs categorizations

Commit:
Merge branch 'ft-file-download-endpoint-165421197' of https://github.com/AndelaOSP/art-backend into ft-file-download-endpoint-165421197
